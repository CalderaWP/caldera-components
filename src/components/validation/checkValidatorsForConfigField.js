/**
 * Check the validators for one field
 *
 * @param {Object} configField The field to check the validator of.
 * @param {Object} fieldValues Optional. Data to pass to validator rule callbacks
 * @return {boolean}
 */
import getValidatorsFromConfigField from "./getValidatorsFromConfigField";

export default function checkValidatorsForConfigField(configField,fieldValues) {
	if( ! getValidatorsFromConfigField(configField).length){
		return true;
	}
	let allRulesPassed = true;
	configField.validators.forEach(conditional => {
		if ('function' === typeof conditional && false === conditional.call(null, fieldValues)) {
			allRulesPassed = false;
			return false;
		}
	});

	return allRulesPassed;
}


