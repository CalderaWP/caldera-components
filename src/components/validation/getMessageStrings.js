import strings from './strings';

/**
 * Get translation strings by language
 *
 * @param {String} lang inputType Type of input to get invalid message for.
 * @param {String|Object} lang Language code or an object containing properly formatted strings to search in.
 *
 * @return {Object}
 */
export const getMessageStrings = (lang, defaultLang = 'en') => {
	if ('string' === typeof  lang) {
		if (strings.hasOwnProperty(lang)) {
			return strings[lang];
		} else {
			return getMessageStrings(defaultLang);
		}
	}
	if ('object' === typeof  lang) {
		return Object.assign(getMessageStrings(defaultLang, 'en'), lang);
	}
};

/**
 * Get message for an invalid field, by field type
 *
 * @param {String} lang inputType Type of input to get invalid message for.
 * @param {String|Object} lang Language code or an object containing properly formatted strings to search in.
 * @param {String} defaultLang Optional. Language code to use for missing strings.
 * @return {String}
 */
export const getMessageStringByType = (inputType, lang, defaultLang = 'en') => {
	const strings = getMessageStrings(lang, defaultLang);
	if (strings.type.hasOwnProperty(inputType)) {
		return strings.type[inputType];
	}
	return strings.defaultMessage;
};

/**
 * Get the required message, by language.
 *
 * @param {String|Object} lang Language code or an object containing properly formatted strings to search in.
 * @param {String} defaultLang Optional. Language code to use for missing strings.
 * @return {String}
 */
export  const getRequiredMessage = (lang, defaultLang = 'en' ) => {
	return getMessageStrings(lang, defaultLang).required;

}
