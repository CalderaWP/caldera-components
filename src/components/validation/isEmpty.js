/**
 * CalderaValidators that check if a value is empty or not
 */
export default {
	/**
	 * Check if different types (numbers, strings, objects, arrays or Maps) are empty
	 *
	 * @param {String|number|Array|object|Map} value
	 * @return {boolean}
	 */
	anything(value){
		if( Array.isArray(value)){
			return this.array(value);
		}
		if( value instanceof Map ){
			return this.map(value);
		}
		switch (typeof value){
			case 'object':
				return this.object(value);
			case 'string':
				return this.string(value);
			default:
				return false;//
		}
		return false;
	},
	/**
	 * Check if value is an empty object
	 *
	 * @param {Object} value The object to check
	 * @return {boolean}
	 */
	object(value){
		for(let key in value) {
			if(value.hasOwnProperty(key))
				return false;
		}
		return true;
	},
	/**
	 * Check if value is an empty array
	 *
	 * @param {Array} value Array to check
	 * @return {boolean}
	 */
	array(value){
		return 0 === value.length;
	},
	/**
	 * Check if value is not an empty string
	 *
	 * @param {String} value String to check
	 * @return {boolean}
	 */
	string(value){
		return !value;
	},
	/**
	 * Check if value is not an empty map
	 *
	 * @param {Map} value The Map to check
	 * @return {boolean}
	 */
	map( value ){
		return 0 === value.size;
	}
}