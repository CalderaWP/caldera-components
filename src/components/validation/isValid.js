import {isEmail,isUrl, isDate} from '@helpdotcom/is';

/**
 * Validators for common types of validation needed
 */
export default {
	/**
	 * Check if a given value is an email address
	 *
	 * @param {String} value Value to check
	 * @return {boolean}
	 */
	email(value){
		return isEmail(value);
	},
	/**
	 * Check if a given value is a valid url
	 *
	 * @param {String} value Value to check
	 * @return {boolean}
	 */
	url(value){
		return isUrl(value);
	},
	/**
	 * Check if a given value is an date.
	 *
	 * Considers Date objects and strings conforming to ISO 8601 standard to be valid.
	 *
	 * @param {String|Date} value Value to check
	 * @return {boolean}
	 */
	date(value){
		return isDate(value);
	},
	/**
	 * Check if a given value is numeric
	 *
	 * @param {String|number} value Value to check
	 * @return {boolean}
	 */
	number(value){
		if( Array.isArray( value ) ){
			return false;
		}
		return ! isNaN(value);
	}
}