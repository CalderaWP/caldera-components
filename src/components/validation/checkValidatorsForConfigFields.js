import checkValidatorsForConfigField from "./checkValidatorsForConfigField";
import getValidatorsFromConfigField from "./getValidatorsFromConfigField";

/**
 * Check validators for a collection of configFields
 *
 * Returns false if any validators fail.
 * Returns true if no validators fail, or there are no validators.
 *
 * @param {Array} configFields The fields to check the validators of.
 * @param {Object} fieldValues Optional. Data to pass to validators rules' callbacks
 * @return {Object}
 */
export default function checkValidatorsForConfigFields(configFields,fieldValues) {
	let results = {};
	Object.values( configFields ).forEach(configField => {
		const rules = getValidatorsFromConfigField(configField);
		let thisResult = true;
		if( rules.length ){
			thisResult = checkValidatorsForConfigField(configField,fieldValues);

		}
		results[configField.ID] = thisResult;
	});
	return results;
}

