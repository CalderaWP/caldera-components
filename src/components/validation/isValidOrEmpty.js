import isEmpty from './isEmpty';
import isValid from './isValid';
/**
 * Validators by type that consider an empty value valid.
 *
 * Combines the isEmpty and isValid validations
 */
export default {
	/**
	 * Check if a given value is an email address or empty
	 *
	 * @param {String} value Value to check
	 * @return {boolean}
	 */
	email(value){
		return isEmpty.anything( value ) || isValid.email(value);
	},
	/**
	 * Check if given value is a string or empty
	 *
	 * @param {String} value Value to check
	 */
	string(value){
		return isEmpty.anything( value ) || isValid.string(value);
	},
	/**
	 * Check if given value is a string or number empty
	 *
	 * @param {String} value Value to check
	 */
	stringOrNumber(value){
		return isEmpty.anything(value) || isValid.stringOrNumber(value);
	},
	/**
	 * Check if a given value is a valid url or empty
	 *
	 * @param {String} value Value to check
	 * @return {boolean}
	 */
	url(value){
		return isEmpty.anything( value ) || isValid.url(value);
	},
	/**
	 * Check if a given value is an date.
	 *
	 * Considers Date objects and strings conforming to ISO 8601 standard to be valid.
	 *
	 * @param {String|Date} value Value to check or empty
	 * @return {boolean}
	 */
	date(value){
		return isEmpty.anything( value ) || isValid.date(value);
	},
	/**
	 * Check if a given value is numeric
	 *
	 * @param {String|number} value Value to check or empty
	 * @return {boolean}
	 */
	number(value){
		return isEmpty.anything( value ) || isValid.number(value);
	}
};