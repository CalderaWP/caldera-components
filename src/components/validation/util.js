import getValidatorsFromConfigField from './getValidatorsFromConfigField';
import isValid from './isValid';
import isValidOrEmpty from './isValidOrEmpty';

/**
 * Find the value of a field by ID
 *
 * Returns null if field is not in collection
 *
 * @param {String} fieldId Field to look for
 * @param {Object} fieldValues Values to look for field in
 * @return {String|Array|boolean|number|null}
 */
export const findFieldValueInFieldValues = (fieldId, fieldValues) => {
	if (fieldValues.hasOwnProperty(fieldId)) {
		return fieldValues[fieldId];
	}

	return null;
};

/**
 * Helper function to curry validator functions
 *
 * Validators expect 1 value, not an object of values, this function solves that.
 *
 * @param {String} fieldId Field to look for
 * @param {Function} validator Validator function
 * @return {function(*=): *}
 */
function curryValidator(fieldId, validator) {
	return (fieldValues) => {
		return validator(findFieldValueInFieldValues(fieldId, fieldValues));
	};
}

/**
 * Create validation function based on type and required
 *
 * @param {Object} configField Field to create validations for
 * @param {Array} validators Current array of validators
 * @param {string} validationType Type of validators to add.
 * @return {Array}
 */
function addValidatorsForType(configField, validators, validationType) {
	if (configField.isRequired) {
		validators.push(
			curryValidator(
				configField.ID, isValid[validationType]
			)
		);
	} else {
		validators.push(
			curryValidator(
				configField.ID, isValidOrEmpty[validationType]
			)
		);
	}

	return validators;
}

/**
 * Adds sensible, automatic validation to fields -- if no validators have been added yet.
 *
 * @param {Object} configField
 * @return {*}
 */
export const addAutomaticValidators = (configField) => {
	let validators = getValidatorsFromConfigField(configField);
	if (0 !== validators.length) {
		return configField;
	}

	switch (configField.type) {
	case 'input':
		if( [
			'email',
			'url',
			'date',
			'number',
		].includes( configField.inputType ) ){
			validators = addValidatorsForType(configField, validators, configField.inputType);
		}else if( 'text' === configField.inputType){
			validators = addValidatorsForType(configField, validators, 'string');

		}

		else{
			validators = addValidatorsForType( configField, validators, 'anything' );
		}

		break;
	case 'select':
		validators = addValidatorsForType( configField, validators, 'stringOrNumber' );
		break;
	}


	return {
		...configField,
		validators
	};


};
