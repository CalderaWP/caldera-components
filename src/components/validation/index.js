/*eslint no-undef: "error"*/
/*eslint-env node*/
import getValidatorsFromConfigField from './getValidatorsFromConfigField';
import checkValidatorsForConfigFields from './checkValidatorsForConfigField';
import checkValidatorsForConfigField from './checkValidatorsForConfigFields';
import messageStrings from './messageStrings';
import isEmpty from './isEmpty';
import isValid from './isValid';
import isValidOrEmpty from './isValidOrEmpty';
//Usage here indicates that this is a cross-cutting concern, & it totally is
//Move to shared module?
import {reduceConfigFieldsToValues} from '../conditional-logic/util';

/**
 * One export for validation system
 */
export default {
	getValidatorsFromConfigField,
	checkValidatorsForConfigFields,
	checkValidatorsForConfigField,
	isEmpty,
	isValid,
	isValidOrEmpty,
	reduceConfigFieldsToValues,
	messageStrings
};