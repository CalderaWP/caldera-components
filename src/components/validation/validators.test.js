import validation from './index';
import getValidatorsFromConfigField from './getValidatorsFromConfigField';
import checkValidatorsForConfigField from './checkValidatorsForConfigField';
import {reduceConfigFieldsToValues} from '../conditional-logic/util';
import checkValidatorsForConfigFields from './checkValidatorsForConfigFields';
import isValid from './isValid';

describe('validation export', () => {
	it('exports getValidatorsFromConfigField', () => {
		expect(typeof validation.getValidatorsFromConfigField).toEqual('function');
	});
	it('exports checkValidatorsForConfigFields', () => {
		expect(typeof validation.checkValidatorsForConfigFields).toEqual('function');
	});
	it('exports checkValidatorsForConfigField', () => {
		expect(typeof validation.checkValidatorsForConfigField).toEqual('function');
	});
});

function validateTrue() {
	return true;
}

describe('getValidatorsFromConfigField', () => {

	const validators = [
		() => {
			return false;
		},
		validateTrue
	];

	const configField = {
		ID: 'fld1234',
		validators: validators
	};


	it('Finds the array', () => {
		expect(getValidatorsFromConfigField(configField)).toEqual(validators);
	});

	it('returns empty array if validators is not set', () => {
		expect(getValidatorsFromConfigField({ID: 'fld1'})).toEqual([]);
	});

	it('returns empty array if validators is not an array', () => {
		expect(getValidatorsFromConfigField({
			ID: 'fld1', validators: {
				validateTrue
			}
		})).toEqual([]);
	});

	describe('checkValidatorsForConfigField', () => {
		it('returns true if no validators', () => {
			expect(checkValidatorsForConfigField({
				ID: 'fld1234',
			})).toEqual(true);

		});
		it('returns true if empty validators', () => {
			expect(checkValidatorsForConfigField({
				validators: []
			})).toEqual(true);
		});

		it('returns true if all validators return true', () => {
			expect(checkValidatorsForConfigField({
				validators: [
					validateTrue()
				]
			})).toEqual(true);
		});

		it('returns false if any validators return false', () => {
			expect(checkValidatorsForConfigField({
				validators: [
					() => {
						return false;
					},
					validateTrue()
				]
			})).toEqual(false);
		});
	});


	describe('checkValidatorsForConfigFields', () => {
		let valuesPassedToValiator = {};
		beforeEach(() => {
			valuesPassedToValiator = {};
		});

		const configFields = [
			{
				type: 'email',
				ID: 'validEmailField',
				validators: [
					() => {
						return true;
					},
				]
			},
			{
				type: 'email',
				ID: 'invalidEmailField',
				validators: [
					() => {
						return false;
					},
				]
			},
			{
				type: 'email',
				ID: 'otherField',
				value: 'pants',
				validators: [
					(fieldValues) => {
						valuesPassedToValiator = fieldValues;
						return true;
					},
				]
			},
		];

		const fieldValues = reduceConfigFieldsToValues(configFields);

		it('passes field values to callbacks', () => {
			checkValidatorsForConfigFields(configFields, fieldValues);
			expect(valuesPassedToValiator).toEqual(fieldValues);
		});

		describe('reports results', () => {
			const results = checkValidatorsForConfigFields(configFields, fieldValues);
			it('reports valid', () => {
				expect(results.validEmailField).toBe(true);
			});
			it('reports invalid', () => {
				expect(results.invalidEmailField).toBe(false);
			});
		});
	});

	describe('using default validators', () => {
		const configFieldsWithValidators = [
			{
				type: 'email',
				ID: 'validEmailField',
				validators: [
					(fieldValues) => {
						return isValid.email(fieldValues.validEmailField);
					},
				],
				value: 'roy@hiroy.club'
			},
			{
				type: 'email',
				ID: 'invalidEmailField',
				validators: [
					(fieldValues) => {
						return isValid.email(fieldValues.invalidEmailField);
					},
				],
				value: 'pants'
			},
			{
				type: 'number',
				ID: 'invalidNumber',
				validators: [
					(fieldValues) => {
						return isValid.number(fieldValues.invalidNumber);
					},
				],
				value: 'pants'
			},
			{
				type: 'number',
				ID: 'validNumber',
				validators: [
					(fieldValues) => {
						return isValid.number(fieldValues.validNumber);
					},
				],
				value: 1
			},
			{
				type: 'date',
				ID: 'validDate',
				validators: [
					(fieldValues) => {
						return isValid.number(fieldValues.validNumber);
					},
				],
				value: new Date('13 October 1982 00:00 UTC')
			},
			{
				type: 'date',
				ID: 'invalidDate',
				validators: [
					(fieldValues) => {
						return isValid.number(fieldValues.invalidDate);
					},
				],
				value: 'Nothing stops Bluma'
			},
			{
				type: 'date',
				ID: 'validDate',
				validators: [
					(fieldValues) => {
						return isValid.number(fieldValues.validNumber);
					},
				],
				value: new Date('13 October 1982 00:00 UTC')
			},
			{
				type: 'url',
				ID: 'invalidUrl',
				validators: [
					(fieldValues) => {
						return isValid.url(fieldValues.invalidUrl);
					},
				],
				value: 'Nothing stops Bluma'
			},
			{
				type: 'url',
				ID: 'validUrl',
				validators: [
					(fieldValues) => {
						return isValid.url(fieldValues.validUrl);
					},
				],
				value: 'https://corkum.mike'
			},

		];
		const fieldValuesForThisTest = reduceConfigFieldsToValues(configFieldsWithValidators);


		it('Produces the right results', () => {
			expect(JSON.stringify(checkValidatorsForConfigFields(configFieldsWithValidators, fieldValuesForThisTest))).toMatchSnapshot();
		});


	});
});

