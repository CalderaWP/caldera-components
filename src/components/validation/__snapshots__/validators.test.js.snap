// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getValidatorsFromConfigField using default validators Produces the right results 1`] = `"{\\"validEmailField\\":true,\\"invalidEmailField\\":false,\\"invalidNumber\\":false,\\"validNumber\\":true,\\"validDate\\":true,\\"invalidDate\\":false,\\"invalidUrl\\":false,\\"validUrl\\":true}"`;
